INT-2354 allow ExecutorService to be injected via constructor

boaty82 opened the pull request:
INT-2354: Add support for ExecutorService injection in LockRegistryLeaderInitiator

https://github.com/spring-projects/spring-integration/issues/2354

* add 2 constructor overloads

boaty82 committed:
INT-2354 allow ExecutorService to be injected via constructor

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/leader/LockRegistryLeaderInitiator.java:
```diff
@@ -162,10 +162,33 @@ public LockRegistryLeaderInitiator(LockRegistry locks) {
+	 * @param candidate leadership election candidate
+	 * @param executorService for running leadership election daemon
+	 */
+	public LockRegistryLeaderInitiator(LockRegistry locks, Candidate candidate, ExecutorService executorService) {
```
I would say this option is not so important.
Let's move it just into a setter!

boaty82 replied:
OK, was just trying to keep the changes to a minimum.

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/support/leader/LockRegistryLeaderInitiatorTests.java:
```diff
@@ -76,17 +78,40 @@ public void init() {
+	}
+
+	@Test
+	public void startAndStopWithRegistryAndExecutorService() throws Exception {
```
I think your new tests don't bring any value.

Would be better to add setter, set an external executor, call `destroy()` on the `LockRegistryLeaderInitiator` and in the end ensure that exactly this external executor is used. Plus its `shutdown()` hasn't been called.

TO repeat a test-case with the same logic doesn't expose any vulnerability.

boaty82 replied:
Fair points, will make the changes and was just trying to cover my changes.  Certainly the destroy() and shutdown() verifies make sense.  will make changes and update PR

artembilan replied:
Good. Thank you very much for the update! I will come to you for review tomorrow. 

For this particular case to check an internal executor you can consider to use `TestUtils.getPropertyValue()` where we use reflection on the matter

boaty82 replied:
Ok awesome. I’ll add the test case and then push, most likely tomorrow

artembilan reviewed (changes requested):
Thank you for contribution, but, please, take a look into my comments.

boaty82 committed:
Revert "INT-2354 allow ExecutorService to be injected via constructor"

This reverts commit 3b358c8

boaty82 commented:
Ok awesome that should work then. I’ll finish off the changes and push tonight/tomorrow

boaty82 committed:
INT-2354 allow ExecutorService to be injected via setter

Changes following feedback on https://github.com/spring-projects/spring-integration/pull/2357

boaty82 committed:
INT-2354 missed file in previous commit

boaty82 commented:
OK, well the first failure was my own fault I did add the change to my staged changes before commit and pushing.

However, the second failure doesn't appear to be anything I've done.  The test failing in CI is not failing for me locally.  Have ran the test suite 3 times and the only issue with a test is

`java.lang.AssertionError: 
Expected: an instance of java.net.UnknownHostException
     but: <java.net.ConnectException: Operation timed out> is a java.net.ConnectException
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at org.springframework.integration.ip.tcp.connection.ConnectionEventTests.testFailConnect(ConnectionEventTests.java:339)
`

but this failed even before I made any changes after forking

artembilan commented:
I think the problem with the `PollingConsumerEndpointTests` that it uses shared mock object in between tests and relies on the `Mockito.reset()`. In other words it doesn’t follow Mockito recommendations. Different story, though, but you feel free to come back to us with the contribution on the matter as well!

boaty82 commented:
OK sure, I can read up on that on the train tomorrow morning (9:15PM now, so going to log off).  But what to do with the current situation?  Can you trigger the build again?

artembilan commented:
The `ConnectionEventTests.testFailConnect()` should check for both exceptions, because we don’t have guarantee which one is thrown on which OS and on what Java version. Seems for me we had similar problem in other tests before. Or we can just use there a mock for connection factory. Different story as well. Not a part of this OR.

artembilan commented:
Hm. I see there is a new problem:
```
org.springframework.integration.redis.leader.RedisLockRegistryLeaderInitiatorTests > testDistributedLeaderElection FAILED
    java.lang.NullPointerException at RedisLockRegistryLeaderInitiatorTests.java:62
```
That’s might be related to your changes...
If not, good to have any way. There is always a room for improvements!

boaty82 commented:
Yes that indeed seems to be related to my change, apologies. I’ll clean, rerun and manually check the entire suite and go from there

boaty82 committed:
INT-2354 fix failing test

was missed due to 'failing fast' approach and ConnectionEventTests failing due to test not correct for Mac OSX

boaty82 commented:
My local test run didn't pick up the Redis class due to a ''fail fast' approach for the build and the before mentioned issue with `ConnectionEventTests.testFailConnect` - however, I should've just simply traced usages of the constructor within the project and gone from there.

Now, looking at `PollingConsumerEndpointTests.heavierLoadTest()`

Firstly, I'm not quite sure what this adds over the existing tests - other than just running the test `multipleMessages` many, many times

Anyway, I believe the solution is quite simple.  In the `init` method, before constructing the endpoint, create a new Mock instance of PollableChannel and then remove the line at end of init
`Mockito.reset(channelMock)`

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/leader/LockRegistryLeaderInitiator.java:
```diff
@@ -168,6 +175,14 @@ public LockRegistryLeaderInitiator(LockRegistry locks, Candidate candidate) {
 	}
 
+	@Override
+	public void afterPropertiesSet() throws Exception {
```
We don't need this method here. We can just simply have that 
```
private ExecutorService executorService =
			Executors.newSingleThreadExecutor(new CustomizableThreadFactory("lock-leadership-"));
```
 as before.

Indeed we need some `boolean` property to control external executor. But in other places we call it like `executorServiceExplicitlySet` - much cleaner and consistent.
You set it to `true` in the setter

boaty82 replied:
OK, I'll go back to that.  I just personally don't like setter doing anything other than setting the target variable.  Also I thought `afterPropertiesSet()` would given more chance for extension and would also not create an `ExecutorService` & `ThreadFactory` when not needed.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/leader/LockRegistryLeaderInitiator.java:
```diff
@@ -183,12 +198,22 @@ public void setBusyWaitMillis(long busyWaitMillis) {
+	/**
+	 * Sets the {@link ExecutorService}
+	 *
+	 * @param executorService the executor service
```
The method JavaDocs should be like: 
- Use an imperative style (i.e. Return and not Returns) for the first sentence.
- No blank lines
- Must be `@since` when it is new public API.

https://github.com/spring-projects/spring-framework/wiki/Code-Style#javadoc-formatting

artembilan commented on spring-integration-jdbc/src/test/java/org/springframework/integration/jdbc/leader/JdbcLockRegistryLeaderInitiatorTests.java:
```diff
@@ -74,6 +74,7 @@ public void testDistributedLeaderElection() throws Exception {
 			LockRegistryLeaderInitiator initiator = new LockRegistryLeaderInitiator(
 					new JdbcLockRegistry(lockRepository),
 					new DefaultCandidate("foo", "bar"));
+			initiator.afterPropertiesSet();
```
When you reinstate the ctor behavior in the `LockRegistryLeaderInitiator`, all these changes in these tests won't be necessary.

artembilan reviewed (changes requested):
Here is a review for your changes.

artembilan commented:
If you would like to fix those non-related tests, please, consider to do that in the separate PRs.
If you don't know how to proceed with them, please, raise issues on the matter and will see how to be.

The main concern here that they are not related to the subject of the PR and therefore such a discussion here is confusing a bit.

Thanks

boaty82 commented:
Sure, I will create another PR to fix the other test issue

boaty82 committed:
INT-2354 changes following feedback

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/leader/LockRegistryLeaderInitiator.java:
```diff
@@ -92,6 +87,17 @@
 	private final Candidate candidate;
 
+	/**
+	 * flag to denote whether the {@link ExecutorService} was provided via the setter and
```
I would say JavaDocs on the properties don't make sense, more over if they don't start from the capital letter 😄

boaty82 replied:
Given that the variable was renamed to `executorServiceExplicitlySet`, based on your feedback, it doesn't really convey any meaning as to its use, hence why I previously called it `shutdownExecutorService` 

Without a more meaningful name I felt that putting a JavaDoc on it to explain its use was wise, even though the code should explain what is doing rather than any commentary

I have capitalised the first letter of the docs, if you don't want any docs on this property then I can remove - if you feel the comment is not appropriate then please provide an example of what you feel would be more appropriate

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/leader/LockRegistryLeaderInitiator.java:
```diff
@@ -168,6 +174,16 @@ public LockRegistryLeaderInitiator(LockRegistry locks, Candidate candidate) {
 	}
 
+	/**
+	 * Sets the {@link ExecutorService}
```
I told already that method JavaDocs should go in the imperative style: https://github.com/spring-projects/spring-integration/pull/2357#discussion_r167612477.

Also would be great to mention what is default.

boaty82 replied:
I was following other setter docs in the class

```
/**
 * Sets the {@link LeaderEventPublisher}.
 * @param leaderEventPublisher the event publisher
 */
```

I will add to description
`where is not provided then a default of single thread Executor will be used`

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/support/leader/LockRegistryLeaderInitiatorTests.java:
```diff
@@ -284,17 +290,41 @@ public void testExceptionFromLock() throws Exception {
+	public void shouldShutdownInternalExecutorService() throws Exception {
+
+		this.initiator.start();
+		this.initiator.destroy();
```
I think there is no guarantee in the method execution order in one test class.
So, this `destroy()` on the shared variable is going to be dangerous when this test method will be called before any other.

Does it make sense to you?

boaty82 replied:
JUnit instantiates an instance of the test class, per test method, and as none of the variables in the test are static there are no variables shared between tests

artembilan replied:
OK. I see what you mean. Good to know 😄

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/support/leader/LockRegistryLeaderInitiatorTests.java:
```diff
@@ -284,17 +290,41 @@ public void testExceptionFromLock() throws Exception {
+		this.initiator.start();
+		this.initiator.destroy();
+
+		ExecutorService executorService = (ExecutorService) TestUtils.getPropertyValue(this.initiator, "executorService");
```
There is an overloaded `getPropertyValue()` with an explicit class and generic argument to return.

boaty82 committed:
INT-2354 feedback changes

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/leader/LockRegistryLeaderInitiator.java:
```diff
@@ -168,6 +174,17 @@ public LockRegistryLeaderInitiator(LockRegistry locks, Candidate candidate) {
 	}
 
+	/**
+	 * Sets the {@link ExecutorService}, where is not provided then a default of
```
You don't listen to me: must be imperative. Therefore `Set`, not `Seets`

> I was following other setter docs in the class

Then the other one is wrong and must be fixed as well.

Thanks

boaty82 replied:
OK I'll update both my doc and the existing one.

I was confused by you saying mine didn't follow convention when then existing didn't either but was already in the source.

artembilan replied:
Yeah... Sorry for confuse.
What I can tell about that code that it was there before obtaining conventions for the JavaDocs.
Or we have just missed it during review 😄 

Good to have you with a perceptive eye to spot and point for that issue 😉 !

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/leader/LockRegistryLeaderInitiator.java:
```diff
@@ -92,6 +87,17 @@
+	 * Flag to denote whether the {@link ExecutorService} was provided via the setter and
+	 * thus should not be shutdown when {@link #destroy()} is called
+	 */
+	private boolean executorServiceExplicitlySet;
```
> Given that the variable was renamed to `executorServiceExplicitlySet`, based on your feedback, it doesn't really convey any meaning as to its use, hence why I previously called it `shutdownExecutorService`.

The `shutdownExecutorService` is more misleading for me. That may mean that there is a special configurator for this property and it really isn't related to the `executorService`.

The `executorServiceExplicitlySet` explicitly tells us that it related to the action around the `executorService` property and cannot be configured separately.

That is really self-explanatory for me. Not sure what is your confuse. But I'm fine to leave the JavaDoc here.
Only the concern is to add a period in the end of the sentence.

boaty82 replied:
>The executorServiceExplicitlySet explicitly tells us that it related to the action around the executorService property and cannot be configured separately.

OK I take your point, but again I am used to self documenting code hence I went for a descriptive name.  

Not a problem, every day is a school day! 😄

boaty82 committed:
INT-2354 feedback changes

artembilan reviewed (approved):
LGTM. Merging...

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/78b80ef79fed817143833a5f0d08b2753966a300.

Thank you for the contribution.
Looking forward for more!
